README.md
# simple-encryptor

A simple encryptor/decryptor for Node.js.

[![NPM](https://nodei.co/npm/simple-encryptor.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/simple-encryptor/)

[![Build Status](https://travis-ci.org/sehrope/node-simple-encryptor.svg?branch=master)](https://travis-ci.org/sehrope/node-simple-encryptor)

# Installation

Add it to your node.js project via:

```sh
npm install simple-encryptor --save
```

# Usage
First create an encryptor:

```js
// Specify a string key:
// Don't do this though, your keys should most likely be stored in env variables
// and accessed via process.env.MY_SECRET_KEY
var key = 'real secret keys should be long and random';

// Create an encryptor:
var encryptor = require('simple-encryptor')(key);
```

To encrypt something:

```js
var encrypted = encryptor.encrypt('testing');
// Should print gibberish:
console.log('encrypted: %s', encrypted);
```

To decrypt it:

```js
var decrypted = encryptor.decrypt(encrypted);
// Should print 'testing'
console.log('decrypted: %s', decrypted);
```

To generate an HMAC:

```js
var myHmac = encryptor.hmac('testing');
```

Encrypt/decrypt an object (not just a string!):

```js
// nested object:
var obj = {
  foo: {
    bar: [1, "baz"]
  }
};
var objEnc = encryptor.encrypt(obj);
// Should print gibberish:
console.log('obj encrypted: %s', objEnc);
var objDec = encryptor.decrypt(objEnc);
// Should print: {"foo":{"bar":[1,"baz"]}}
console.log('obj decrypted: %j', objDec);
```

# Features

* Encrypt arbitrary objects, not just strings (objects are converted to/from JSON)
* Unique IV per call so no two calls should return the same result for the same input
* Defaults to encrypt-then-mac with AES-256-CBC and SHA-256 HMAC
* Optionally disable HMACs for shorter results
* No complicated options
* Defaults to rejecting short keys (min length is 16)
* Written to be easy to read

# API
The module provides three functions:

* `encryptor.encrypt(obj)` - Encrypt the object and return back the encrypted cipher text. The object is first converted to text via `JSON.stringify(...)`. This means you can encrypt arbitrary objects.
* `encryptor.decrypt(cipherText)` - Decrypts the cipher text and returns the original object. Specifically, it decrypts the cipher text and calls `JSON.parse(...)` on the result. If an error occurs during the decryption then null is returned. __Note:__ This function never throws an error for bad input, it just returns `null`.
* `encryptor.hmac(string)` - Calculate the HMAC of the input string.

# Options
This module supports two forms of creating an encryptor:

## String Key - `encryptor(key)`
If the first parameter is a string then it will be used as the key and the rest of the options will be defaulted.

Example:

```js
// Don't hard code keys! They should be in environment variables!
var encryptor = require('simple-encryptor')('my secret key');
```

## Options hash - `encryptor(opts)`
Alternatively you can specify the string key and other options as a hash. The following properties are supported:

* `key` - the string key to derive the crypto key from. Specifically the crypto key will be derived as the SHA-256 hash of this key. This must be specified, there is no default.
* `hmac` - whether or not to calculate the HMAC of the encrypted text and add that to the result. Additionally, if enabled this will verify the HMAC prior to decrypting. Adding HMACs will add 64-bytes to the result of each encryption (32-byte HMAC stored as hex). By default this is true.
* `reviver` - you can pass in a custom [reviver](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter) function that will be used during decryption. Useful, for example, when your payload contains a date object and you want it to be recreated during decryption.
* `debug` - whether to log errors decrypting, by default this is false.

Example:

```js
// Don't hard code keys! They should be in environment variables!
var encryptor = require('simple-encryptor')({
  key: 'my secret key',
  hmac: false,
  debug: true
});
```

# Internals
Interally this module uses the node.js crypto package. Specifically it uses the specified string key to derive a key via computing it's SHA-256 hash. Encryption is done via AES-256 with a unique IV (intialization vector) per call that is returned as part of the result.

# Generating a key
If you're on a *nix system then the easiest way to generate a random string for a crypto key is to use /dev/urandom. The following will print out 32 random characters of lower case letters, upper case letters, and numbers:

```sh
$ echo "$(< /dev/urandom tr -dc A-Za-z0-9 | head -c 32)"
```

# Dependencies
[scmp](https://www.npmjs.org/package/scmp) for constant-time string comparison.

# License
This plugin is released under the MIT license. See the file [LICENSE](LICENSE).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [4.0.0] - 2020-05-30
### Breaking Changes
- Changed to export a non-default function createEncryptor(...) and update the typings to match.

## [3.0.0] - 2019-04-29
### Changed
- Added engines with min version of node v4.5.0
- Changed "aes256" cipher alias to full name of "aes-256-cbc"
### Breaking Changes
- Changed to use Buffer.from(...) which requires node >= v4.5.0

## [2.0.0] - 2018-08-30
### Added
- This changelog to track future changes.
### Changed
- Export format for TypeScript type definitions updated to use "export default ..." style.
### Breaking Changes
- None for core library vs v1.x but the type definition changes may cause compile errors.


test/index.js
var assert = require('assert');

var testKey = 'real secret keys should be long and random';
var otherKey = 'some key that is different than testKey';

function testEncryptor(encryptor) {
  function testRoundTrip(val) {
    var encrypted = encryptor.encrypt(val);
    var decrypted = encryptor.decrypt(encrypted);
    assert.deepEqual(decrypted, val);
  }

  it('should encrypt and decrypt text', function() {
    testRoundTrip('testing');
  });

  it('should encrypt and decrypt numbers', function() {
    testRoundTrip(12345);
  });

  it('should encrypt and decrypt null', function() {
    testRoundTrip(null);
  });

  it('should encrypt and decrypt nested objects', function() {
    var obj = {
      foo: {
        bar: {
          baz: "some text",
          quz: 12345
        }
      }
    };
    var encrypted = encryptor.encrypt(obj);
    var decrypted = encryptor.decrypt(encrypted);
    assert.equal(JSON.stringify(decrypted), JSON.stringify(obj));
  });

  it('should generate unique output for encrypt', function() {
    var cache = {};
    for(var i=0;i<100;i++) {
      var encrypted = encryptor.encrypt('this is a test');
      assert(!cache[encrypted]);
      cache[encrypted] = true;
    }
  });

  // NOTE: This test will print 'Exception in decrypt...' log lines
  it('should return null when decrypted with the wrong key', function() {
    var otherEncryptor = require('../index')(otherKey);
    var encrypted = otherEncryptor.encrypt('testing');
    var decrypted = encryptor.decrypt(encrypted);
    assert.equal(decrypted, null);
  });
};

describe('An encryptor with a short key', function() {
  it('should not create the encryptor', function() {
    assert.throws(function() {
      var encryptor = require('../index')('short key');
    });
  });
});

describe('An encryptor with the defaults', function() {
  var encryptor = require('../index')(testKey);
  testEncryptor(encryptor);
});

describe('An encryptor with HMACs disabled', function() {
  var opts = {
    key: testKey,
    hmac: false
  }
  var encryptor = require('../index')(opts);
  testEncryptor(encryptor);
});

describe('An encryptor with custom reviver', function() {
  // parses ISO 8601 date strings
  function dateReviver(key, value) {
    var a;
    if (typeof value === 'string') {
      a = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}).(\d{3})Z$/.exec(value);
      if (a) {
        return new Date(Date.UTC(+a[1], +a[2] - 1, +a[3], +a[4], +a[5], +a[6], +a[7]));
      }
    }
    return value;
  };

  var opts = {
    key: testKey,
    reviver: dateReviver
  }
  var encryptor = require('../index')(opts);
  it('should encrypt and decrypt dates with a date reviver', function() {
    var obj = {
      foo: new Date()
    };
    var encrypted = encryptor.encrypt(obj);
    var decrypted = encryptor.decrypt(encrypted);
    assert.equal(JSON.stringify(decrypted), JSON.stringify(obj));
  });

  it('should throw an error if the reviver is not a function', function() {
    assert.throws(function() {
      var encryptor = require('../index')({key: testKey, reviver: 'not-a-function'});
    });
  });
});


test/vectors.js
var assert = require('assert');
var createEncryptor = require('../index');

describe('An encrypter with a key of abcd1234abcd1234', () => {
    const key = 'abcd1234abcd1234';
    const encrypter = createEncryptor(key);

    it('should decrypt the previously encrypted string', () => {
        const cipherText = '6e558a97efc5da4203bdbbc33ff038158296dce37eb7de77914799b5e30190db03515c846e4a33f85282130aa7f52ea0n+6fPdrB34sxg322vg9ERA==';
        const expectedPlainText = 'test';
        const plainText = encrypter.decrypt(cipherText);
        assert.equal(plainText, expectedPlainText);
    });

    it('should decrypt the previously encrypted object', () => {
        const cipherText = 'e62134f0d701b616342d2a0f96fd79b3eb0f33302db49661ed143d7b3a9c0474d28a0bd97b745d20876ff3944ac3adf2Yn1up3jR3MBUeH1NHsj5ow==';
        const expectedPlainText = {
            foo: 123,
        };
        const plainText = encrypter.decrypt(cipherText);
        assert.deepEqual(plainText, expectedPlainText);
    });

    it('should decrypt the previously encrypted nested object', () => {
        const cipherText = '4f9d01cc50216fcfeba82ae1894f70b5ff19d2ea2bb16b6d9f717aca56944ecd7efbf82ed0fcc210158a91830289f0bcmRuJHzJOUxgiFc9PT8orrycagEKpf/uS86rKYLheRBo=';
        const expectedPlainText = {
            foo: [
                {
                    bar: 123,
                },
            ],
        };
        const plainText = encrypter.decrypt(cipherText);
        assert.deepEqual(plainText, expectedPlainText);
    });
});


index.d.ts
export interface SimpleEncryptor {
    /**
     *  Encrypts an arbitrary object using the derived cryptoKey and retursn the result as text.
     *  The object is first serialized to JSON (via JSON.stringify) and the result is encrypted.
     */
    encrypt(obj: any): string

    /**
     *  Decrypts the encrypted cipherText and returns back the original object deserialized from JSON.
     *  If the cipherText cannot be decrypted (bad key, bad text, bad serialization) then it returns null.
     */
    decrypt<T = any>(cipherText: string): T | null;

    /**
     *  Returns the HMAC(text) using the derived cryptoKey
     *  Defaults to returning the result as hex.
     */
    hmac(text: string, encoding?: string): string
}

export interface SimpleEncryptorOptions {
    key: string;
    hmac: boolean;
    debug: boolean;
    reviver?(key: any, value: any): any;
}

export function createEncryptor(opts: SimpleEncryptorOptions): SimpleEncryptor;
export function createEncryptor(key: string): SimpleEncryptor;


index.js
var crypto = require('crypto');
var scmp = require('scmp');

// Arbitrary min length, nothing should shorter than this:
var MIN_KEY_LENGTH = 16;

function createEncryptor(opts) {
  if( typeof(opts) == 'string' ) {
    opts = {
      key: opts,
      hmac: true,
      debug: false
    };
  }
  var key = opts.key;
  var verifyHmac = opts.hmac;
  var debug = opts.debug;
  var reviver = opts.reviver;

  if( !key || typeof(key) != 'string' ) {
    throw new Error('a string key must be specified');
  }
  if( key.length < MIN_KEY_LENGTH ) {
    throw new Error('key must be at least ' + MIN_KEY_LENGTH + ' characters long');
  }
  if( reviver !== undefined && reviver !== null && typeof(reviver) != 'function' ) {
    throw new Error('reviver must be a function');
  }

  // Use SHA-256 to derive a 32-byte key from the specified string.
  // NOTE: We could alternatively do some kind of key stretching here.
  var cryptoKey = crypto.createHash('sha256').update(key).digest();

  // Returns the HMAC(text) using the derived cryptoKey
  // Defaults to returning the result as hex.
  function hmac(text, format) {
    format = format || 'hex';
    return crypto.createHmac('sha256', cryptoKey).update(text).digest(format);
  }

  // Encrypts an arbitrary object using the derived cryptoKey and retursn the result as text.
  // The object is first serialized to JSON (via JSON.stringify) and the result is encrypted.
  //
  // The format of the output is:
  // [<hmac>]<iv><encryptedJson>
  //
  // <hmac>             : Optional HMAC
  // <iv>               : Randomly generated initailization vector
  // <encryptedJson>    : The encrypted object
  function encrypt(obj) {
    var json = JSON.stringify(obj);

    // First generate a random IV.
    // AES-256 IV size is sixteen bytes:
    var iv = crypto.randomBytes(16);

    // Make sure to use the 'iv' variant when creating the cipher object:
    var cipher = crypto.createCipheriv('aes-256-cbc', cryptoKey, iv);

    // Generate the encrypted json:
    var encryptedJson = cipher.update(json, 'utf8', 'base64') + cipher.final('base64');

    // Include the hex-encoded IV + the encrypted base64 data
    // NOTE: We're using hex for encoding the IV to ensure that it's of constant length.
    var result = iv.toString('hex') + encryptedJson;

    if( verifyHmac ) {
      // Prepend an HMAC to the result to verify it's integrity prior to decrypting.
      // NOTE: We're using hex for encoding the hmac to ensure that it's of constant length
      result = hmac(result, 'hex') + result;
    }

    return result;
  }

  // Decrypts the encrypted cipherText and returns back the original object.
  // If the cipherText cannot be decrypted (bad key, bad text, bad serialization) then it returns null.
  //
  // NOTE: This function never throws an error. It will instead return null if it cannot decrypt the cipherText.
  // NOTE: It's possible that the data decrypted is null (since it's valid input for encrypt(...)).
  //       It's up to the caller to decide if the result is valid.
  function decrypt(cipherText) {
    if( !cipherText ) {
      return null;
    }
    try {
      if( verifyHmac ) {
        // Extract the HMAC from the start of the message:
        var expectedHmac = cipherText.substring(0, 64);
        // The remaining message is the IV + encrypted message:
        cipherText = cipherText.substring(64);
        // Calculate the actual HMAC of the message:
        var actualHmac = hmac(cipherText);
        if( !scmp(Buffer.from(actualHmac, 'hex'), Buffer.from(expectedHmac, 'hex')) ) {
          throw new Error('HMAC does not match');
        }
      }

      // Extract the IV from the beginning of the message:
      var iv = Buffer.from(cipherText.substring(0,32), 'hex');
      // The remaining text is the encrypted JSON:
      var encryptedJson = cipherText.substring(32);

      // Make sure to use the 'iv' variant when creating the decipher object:
      var decipher = crypto.createDecipheriv('aes-256-cbc', cryptoKey, iv);
      // Decrypt the JSON:
      var json = decipher.update(encryptedJson, 'base64', 'utf8') + decipher.final('utf8');

      // Return the parsed object:
      return JSON.parse(json, reviver);
    } catch( e ) {
      // If we get an error log it and ignore it. Decrypting should never fail.
      if( debug ) {
        console.error('Exception in decrypt (ignored): %s', e);
      }
      return null;
    }
  }

  return {
    encrypt: encrypt,
    decrypt: decrypt,
    hmac: hmac
  };
}

module.exports = createEncryptor;
module.exports.createEncryptor = createEncryptor;
